### DISCUSSION TITLE
Command Error Event

### DISCUSSION THREAD

**Discussion by IIDarkNessYT** [Rapptz/discord.py#8224] [Q&A] [answered]
Help me please! I am doing slash commands in the bot, and I need to make an event on command error, as well as recognize the error (for example, when a person has no rights (Missing Permissions)). Please tell me the code!

Update: I'm use slash-commands (app commands)

**Comment by LostLuma** [comment, marked as answer]
You can attach an error handler to the `CommandTree` of your client by either using the [`error`](https://discordpy.readthedocs.io/en/latest/interactions/api.html#discord.app_commands.CommandTree.error) decorator or subclassing `CommandTree` and overwriting the [`on_error`](https://discordpy.readthedocs.io/en/latest/interactions/api.html#discord.app_commands.CommandTree.on_error) callback. Here is an example of the former:

```py
client = discord.Client(...)
tree = discord.app_commands.CommandTree(client)


@tree.error
async def on_app_command_error(interaction: discord.Interaction, error: discord.app_commands.AppCommandError) -> None:
    ...
```

You can check which exact subclass of `AppCommandError` was raised here to handle it, they are listed in the [Interactions API Documentation](https://discordpy.readthedocs.io/en/latest/interactions/api.html#exceptions).

**Reply by IIDarkNessYT** [reply]
Thank you!

**Reply by M9-SD** [reply]
here is an example that handles app_commands.checks.cooldown
```
@tree.error
async def on_app_command_error(interaction: discord.Interaction, error: discord.app_commands.AppCommandError) -> None:
    if isinstance(error, app_commands.errors.CommandOnCooldown):
        await interaction.response.send_message(f'Command "{interaction.command.name}" is on cooldown, you can use it in {round(error.retry_after, 2)} seconds.', ephemeral=True)
```

**Reply by Soheab** [reply]
> here is an example that handles app_commands.checks.cooldown
> 
> ```python
> @tree.error
> async def on_app_command_error(interaction: discord.Interaction, error: discord.app_commands.AppCommandError) -> None:
>     if isinstance(error, app_commands.errors.CommandOnCooldown):
>         await interaction.response.send_message(f'Command "{interaction.command.name}" is on cooldown, you can use it in {round(error.retry_after, 2)} seconds.', ephemeral=True)
> ```

I would not recommend ignoring all exceptions except `app_commands.CommandOnCooldown` even if it's an example, it's a common problem in the dpy help channels where users copy any code to see and especially from an answered discussion.
